src/vsr/replica.zig

```diff
@@ -4899,112 +4948,322 @@ pub fn ReplicaType(
+            // replica 0 doesn't have 6/7, then replica 1/2 must share the latest view_normal. ∎)
+
+            // Now that the canonical headers are all in place, repair any other headers:
+            for (self.do_view_change_from_all_replicas) |received| {
+                if (received) |message| {
+                    const view_normal = @intCast(u32, message.header.timestamp);
+                    assert(view_normal < message.header.view);
+
+                    if (view_normal == view_normal_canonical) continue;
+                    assert(view_normal < view_normal_canonical);
+
+                    for (message_body_as_headers(message)) |*header| {
+                        // We must trust headers that other replicas have committed, because
+                        // repair_header() will not repair a header if the hash chain has a gap.
+                        if (header.op <= message.header.commit) {
+                            log.debug(
+                                "{}: on_do_view_change: committed: replica={} op={} checksum={}",
+                                .{
+                                    self.replica,
+                                    message.header.replica,
+                                    header.op,
+                                    header.checksum,
+                                },
+                            );
+                            self.replace_header(header);
+                        } else {
+                            _ = self.repair_header(header);
+                        }
+                    }
+                }
+            }
+
+            // The `prepare_timestamp` prevents a primary's own clock from running backwards.
+            // Therefore, `prepare_timestamp`:
+            // 1. is advanced if behind the cluster, but never reset if ahead of the cluster, i.e.
+            // 2. may not always reflect the timestamp of the latest prepared op, and
+            // 3. should be advanced before discarding the timestamps of any uncommitted headers.
+            const head = self.journal.header_with_op(self.op).?;
+            if (self.state_machine.prepare_timestamp < head.timestamp) {
+                self.state_machine.prepare_timestamp = head.timestamp;
+            }
+
+            // Any uncanonical ops remaining either:
+            // * Connect to the hash chain on the right.
+            // * Do not connect on the right (hash chain break).
+            //
+            // If there is a hash chain break, none of the headers from the canonical DVCs replaced
+            // the broken op. It is truncated like a gap.
+            //
+            // Removing these is necessary for correctness and liveness, to ensure that
+            // disconnected headers do not remain in place in lieu of gaps.
+            var op: u64 = op_uncanonical - 1;
+            while (op <= self.op) : (op += 1) {
```

Comment:
Can we refactor out some blocks within this function to keep the overall function body under 70 lines?

Reply:
I will refactor this some, but I don't think it will be under 70 lines yet. I have some other changes in mind for a future PR (to more aggressively verify the "all same-view_normal-DVC headers must match" invariant) that will take it the rest of the way. FWIW, I don't think this function actually does "too much" — rather, it has lots of documentation, checks, and logging, which inflates the line count. (Plus we have iterate the DVCs twice, both times filtering+checking `view_normal`).

One (possibly controversial :wink:) refactor — I restructured `discard_uncommitted_ops`:
- Renamed to `do_view_change_op_max`
- It now just returns the new `replica.op`; it does not set `replica.op` or truncate entries from the log.
- I moved the "search for hash break in uncanonical headers" into there as well.

My goal is to move as much of the state mutation as possible directly into `set_log_from_do_view_change_messages`, rather than the replica functions it invokes. (I find having the mutation as close together as possible makes it easier to mentally/manually check invariants).

Reply:
> My goal is to move as much of the state mutation as possible directly into set_log_from_do_view_change_messages, rather than the replica functions it invokes.

Excellent idea. Let's do it.